elf2uf2/elf2uf2.cpp

```diff
@@ -11,49 +11,93 @@
 }
 
+/**
+ * @brief Terminates the program with a write error.
+ */
 static void fail_write_error() {
     fail(ERROR_WRITE_FAILED, "Failed to write output file");
 }
 
+/**
+ * @struct page_fragment
+ * @brief Represents a fragment of a page with file offset and size.
+ */
 struct page_fragment {
-    page_fragment(uint32_t file_offset, uint32_t page_offset, uint32_t bytes) : file_offset(file_offset), page_offset(page_offset), bytes(bytes) {}
-    uint32_t file_offset;
-    uint32_t page_offset;
-    uint32_t bytes;
+    /**
+     * @brief Constructor for page_fragment.
+     * @param file_offset Offset in the input file.
+     * @param page_offset Offset within the page.
+     * @param bytes Number of bytes in the fragment.
+     */
+    page_fragment(uint32_t file_offset, uint32_t page_offset, uint32_t bytes)
+        : file_offset(file_offset), page_offset(page_offset), bytes(bytes) {}
+
+    uint32_t file_offset;  ///< Offset in the input file.
+    uint32_t page_offset;  ///< Offset within the page.
+    uint32_t bytes;        ///< Number of bytes in the fragment.
 };
 
+/**
+ * @brief Checks if an address is within the valid ranges.
+ * 
+ * @param valid_ranges Valid address ranges.
+ * @param addr Physical address to check.
+ * @param vaddr Virtual address.
+ * @param size Size of the area to check.
+ * @param uninitialized Flag indicating if the area is uninitialized.
+ * @param ar Reference to an address_range that will be set on success.
+ * @return 0 on success, error code otherwise.
+ */
 int check_address_range(const address_ranges& valid_ranges, uint32_t addr, uint32_t vaddr, uint32_t size, bool uninitialized, address_range &ar) {
     for(const auto& range : valid_ranges) {
         if (range.from <= addr && range.to >= addr + size) {
             if (range.type == address_range::type::NO_CONTENTS && !uninitialized) {
-                fail(ERROR_INCOMPATIBLE, "ELF contains memory contents for uninitialized memory at %p", addr);
+                fail(ERROR_INCOMPATIBLE, "ELF contains memory contents for uninitialized memory at 0x%08x", addr);
             }
             ar = range;
             if (verbose) {
-                printf("%s segment %08x->%08x (%08x->%08x)\n", uninitialized ? "Uninitialized" : "Mapped", addr,
-                   addr + size, vaddr, vaddr+size);
+                printf("%s segment 0x%08x->0x%08x (0x%08x->0x%08x)\n", 
+                       uninitialized ? "Uninitialized" : "Mapped", 
+                       addr, addr + size, vaddr, vaddr + size);
             }
             return 0;
         }
     }
-    fail(ERROR_INCOMPATIBLE, "Memory segment %08x->%08x is outside of valid address range for device", addr, addr+size);
+    fail(ERROR_INCOMPATIBLE, "Memory segment 0x%08x->0x%08x is outside of valid address range for device", addr, addr + size);
     return ERROR_INCOMPATIBLE;
 }
 
+/**
+ * @brief Checks ELF-32 program header entries and populates the page fragments map.
```

Comment:
@will-v-pi Is this hyphenation of "ELF-32" correct? :man_shrugging:

Reply:
I think it should be un-hyphenated, so ELF32